File: api/normalised-feed.py
from api._lib.blob_io import load_df
from api._lib.handler_utils import BaseHandler, get_processed_df, records


class handler(BaseHandler):
    def do_POST(self):
        try:
            body = self._json_body()
            raw_df = load_df(body["rawDfUrl"])
            params = {**body.get("state", {}), "colourMapBlobUrl": body.get("colourMapBlobUrl")}
            df = get_processed_df(raw_df, params)

            self._send_json(200, {
                "preview": records(df),
                "rowCount": len(df),
                "colCount": len(df.columns),
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

File: api/normalised-feed.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import { sendError, records } from '../api/_lib/handlerUtils.js'
import { loadRows } from '../api/_lib/sessionStore.js'
import { getProcessedRows } from '../api/_lib/pipeline.js'

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' })

  try {
    const { rawDfUrl, state = {} } = req.body as { rawDfUrl: string; state?: Record<string, unknown> }
    const raw = await loadRows(rawDfUrl)
    const processed = await getProcessedRows(raw, state as never)

    res.status(200).json({
      preview: records(processed),
      rowCount: processed.length,
      colCount: Object.keys(processed[0] ?? {}).length,
    })
  } catch (err) {
    sendError(res, err)
  }
}
